tests/test_readers.py

```diff
@@ -182,18 +182,29 @@ def checkExposureFitsReader(self, exposureIn, fileName, dtypesOut):
         self.assertEqual(exposureIn.getMetadata().toDict(), reader.readMetadata().toDict())
         self.assertWcsAlmostEqualOverBBox(exposureIn.getWcs(), reader.readWcs(), self.bbox,
                                           maxDiffPix=0, maxDiffSky=0*degrees)
+        self.assertWcsAlmostEqualOverBBox(exposureIn.getWcs(), reader.readComponent('SKYWCS'),
+                                          self.bbox,
+                                          maxDiffPix=0, maxDiffSky=0*degrees)
         self.assertEqual(exposureIn.getFilter(), reader.readFilter())
         self.assertEqual(exposureIn.getFilterLabel(), reader.readFilterLabel())
+        self.assertEqual(exposureIn.getFilterLabel(), reader.readComponent('FILTER'))
         self.assertEqual(exposureIn.getPhotoCalib(), reader.readPhotoCalib())
+        self.assertEqual(exposureIn.getPhotoCalib(), reader.readComponent('PHOTOCALIB'))
         self.assertImagesEqual(exposureIn.getPsf().computeImage(), reader.readPsf().computeImage())
+        self.assertImagesEqual(exposureIn.getPsf().computeImage(),
+                               reader.readComponent('PSF').computeImage())
         self.assertEqual(exposureIn.getInfo().getValidPolygon(), reader.readValidPolygon())
+        self.assertEqual(exposureIn.getInfo().getValidPolygon(), reader.readComponent('VALID_POLYGON'))
         self.assertCountEqual(exposureIn.getInfo().getApCorrMap(), reader.readApCorrMap())
+        self.assertCountEqual(exposureIn.getInfo().getApCorrMap(), reader.readComponent('AP_CORR_MAP'))
         self.assertEqual(exposureIn.getInfo().getVisitInfo().getExposureTime(),
                          reader.readVisitInfo().getExposureTime())
         point = Point2D(2.3, 3.1)
         wavelengths = np.linspace(4000, 5000, 5)
         self.assertFloatsEqual(exposureIn.getInfo().getTransmissionCurve().sampleAt(point, wavelengths),
                                reader.readTransmissionCurve().sampleAt(point, wavelengths))
+        # Note: readComponent('TRANSMISSION_CURVE') does not work.
```

Comment:
Nothing obviously jumps out at me; it should behave the same as every other component. Are you getting `None`, or some kind of error?

The only possibility I can think of is that the logic in https://github.com/lsst/afw/blob/master/src/image/ExposureFitsReader.cc#L360-L363 is failling. Can you manually inspect the file and see if it has an `ARCHIVE_ID_TRANSMISSION_CURVE` keyword, and that it matches the old-style `TRANSMISSION_CURVE_ID`?

Reply:
It returns a generic `Storable` that hasn't been casted to a `TransmissionCurve`.

Reply:
I think that means there's a bug in `TransmissionCurve`'s pybind11 wrapper. The formal C++ type is `Storable`, but that shouldn't matter in Python.

Reply:
I don't know how this works, and looking at the `transmissionCurve.cc` it seems okay to me compared to others like it.  Any hints greatly appreciated!

Reply:
The problem is that many of the `TransmissionCurve` subclasses are defined in the .cc file (https://github.com/lsst/afw/blob/master/src/image/TransmissionCurve.cc#L58) and are never declared to pybind11, because their interfaces don't extend beyond that of the base class, and users are only supposed to construct them via public factory methods.  The sad thing is that this is a very _good_ pattern for C++ alone (including, I think, in this case), but pybind11 just doesn't support it well : it will only automatically downcast from base (e.g. `Storable`) to most-derived (e.g. `ProductTransmissionCurve`) if the latter is known to it, not some intermediate base class it knows about (e.g. `TransmissionCurve`).  Also bums me out that this is something Boost.Python could do, and pybind11 just thought it was okay to drop it when they set out to replace Boost.Python.

Anyhow, I'm not sure where that leaves us.  I think I'd just continue to special-case everything we have so far in the formatter, and maybe add some warnings in the docs (`TransmissionCurve` may not be the only component affected - new subclasses of other components that aren't declared to pybind11 will suffer the same fate).  We could refactor `TransmissionCurve` itself to avoid the immediate problem, but that feels like a step backward for the C++ interface that doesn't really address the problem anyway.

Reply:
Ah, that would explain it. I was looking at the docs and didn't realize `TransmissionCurve` has subclasses.

Unfortunately, aside from including a special case (perhaps in the pybind11 wrapper for `ExposureFitsReader`) or registering the subclasses with pybind11 I don't see an obvious solution. My proposal to make `readComponent` take a `Key` won't solve this problem, because the pybind11 wrapper would still be working with strings and `Storable`.

Reply:
Ah, I see.  This is rather unfortunate.  Where does this go into the docs?

Reply:
How about a `Notes` section for this `readComponent` and the class docs for `Storable`?  If there's anywhere else @jmeyers314 wrote instructions for trampolining, I'd put it there, too.

Reply:
One can add strings to the pybind11 wrapper `.def` call itself as one of the arguments, but you're right that we don't do that regularly (it'd involve a lot of duplication or code generation), and I was thinking we'd just add it to the C++ Doxygen docs.  I'm _hoping_ the Doxygen/Sphinx integration that's in the works will at least associate those docstrings with the right Python classes/methods, though I have no idea how it could possibly handle non-trivial signature mappings.

Reply:
Is this where the sphinx docs are? https://github.com/lsst/afw/blob/master/src/image/ExposureFitsReader.cc#L393-L404

Reply:
If you mean the Doxygen docs, then yes.

Reply:
I'd also add an implementation note to `TransmissionCurve` that says most objects are of a subclass, and that this may affect pybind11 behavior. It may, as @TallJimbo says, be a good pattern in general, but it's very unusual in our code.

Reply:
I added https://github.com/lsst/afw/pull/558/commits/53cd70ac15c11308f8797bde5728ecdc097fcef5 for the `readComponent` note, but I need explicit direction on these other doc edits since it's further in the weeds of c++ that I don't have the vocabulary/knowledge to describe what is going on here.

Reply:
I'm not sure what to say for `Storable` (it's not really an issue specific to that class), but I can push a blurb for `TransmissionCurve`.

Reply:
That would be awesome, thanks!  (And I agree, this isn't a `Storable` issue per-se, though I can see that the doc string needs to be modified to say that these pybind11 wrappers are non-trivial in some cases?)